pyfr/plugins/ascent.py

```diff
@@ -0,0 +1,441 @@
+            nsvpts = len(svpts)
+
+            mesh_op = shape.sbasis.nodal_basis_at(svpts).astype(self.dtype)
+            soln_op = shape.ubasis.nodal_basis_at(svpts).astype(self.dtype)
+            self.ops[ename] = (mesh_op, soln_op)
+
+    def _eval_exprs(self, intg):
+        # Get the primitive variable names
+        pnames = self.elementscls.privarmap[self.ndims]
+
+        # Compute the gradients
+        if self._gradpinfo:
+            grad_soln = intg.grad_soln
+
+        # Iterate over each element type in the simulation
+        for idx, ename, rgn, mesh_op, soln_op in self._ele_regions_lin:
+            # Subset and transpose the solution
+            soln = intg.soln[idx][..., rgn].swapaxes(0, 1)
+
+            d_str = f'domain_{self.dom_0 + idx}'
+
+            # Interpolate the solution to the subdivided points
+            soln = soln_op @ soln
+
+            # Convert from conservative to primitive variables
+            psolns = self.elementscls.con_to_pri(soln, self.cfg)
+
+            # Prepare the substitutions dictionary
+            subs = dict(zip(pnames, psolns), t=intg.tcurr)
+
+            # Prepare any required gradients
+            if self._gradpinfo:
+                grads = np.rollaxis(grad_soln[idx], 2)[..., rgn]
+
+                # Interpolate the gradients to the points
+                grad_soln = soln_op @ grad_soln
+
+                # Transform from conservative to primitive gradients
+                pgrads = self.elementscls.grad_con_to_pri(soln, grads,
+                                                          self.cfg)
+
+                # Add them to the substitutions dictionary
+                for pname, idx in self._gradpinfo:
+                    for dim, grad in zip('xyz', pgrads[idx]):
+                        subs[f'grad_{pname}_{dim}'] = grad
+
+            for field, expr in self._expr.items():
+                if not isinstance(expr, dict):
+                    # Evaluate the expression at each point
+                    fun = npeval(expr, subs).T
+                    self.mesh_n[f'{d_str}/fields/{field}/values'] = fun
+                else:
+                    for sf, sexpr in expr.items():
+                        fun = npeval(sexpr, subs).T
+                        path = f'{d_str}/fields/{field}/values/{sf}'
+                        self.mesh_n[path] = fun
+
+    def _get_expr(self, expr, subs={}):
+        # Ensure the expression does not contain invalid characters
+        if not re.match(r'[A-Za-z0-9_ \t\n\r.,+\-*/%()]+$', expr):
+            raise ValueError('Invalid characters in expression')
+
+        # Substitute variables
+        if subs:
+            expr = re.sub(r'\b({0})\b'.format('|'.join(subs)),
+                          lambda m: str(subs[m[1]]), expr)
+
+        return expr
+
+    def _get_shape(self, name, nspts):
+        shapecls = subclass_where(BaseShape, name=name)
+        return shapecls(nspts, self.cfg)
+    
+    def _init_gradients(self, intg):
```

Comment:
Why are we passing intg?

Reply:
Becasue I copied this code from integrate.py and it did. I'll cleanup, but it might be worth casting this eye over the other plugins as some of the issues seem to be coming from when I have copied code from other places in pyfr that are doing the same but aren't up to the required standard.